File: HC2_lib_python3/HC2.py
#Enoch Chau 2019
#main library for HC2 serial commands for use with Raspberry Pi
import time
import serial
import ac3k_datetime
import math
import csv

class HC2_ROASCII:
    def __init__(self, serial_port):
        self.serial_port = serial_port
        self.id = self.get_id()
        self.rs485_addr = self.get_rs485_addr()
        self.serial_num = self.get_serial_num()


    #Request RDD to get init values
    def write_rdd(self,id,rs485_addr):
        # print("reading values")
        pencil = "{" + str(id) + "99RDD}\r\n"
        # print(pencil)
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        # print(output)

        return output

    #initilization functions
    def get_rs485_addr(self):
        output = self.write_rdd(self.id,99)
        print("rs485 address: " + output[4:6])
        return output[4:6]

    def get_serial_num(self):
        output = self.write_rdd(self.id,self.rs485_addr)
        print("serial number: " + output[83:93])
        return output[83:93]

    def get_id(self):
        alphabet = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
        for id_check in alphabet:
            output = self.write_rdd(str(id_check),99)
            # print(id_check)
            if len(output) > 2:
                break
        print("ID: " + output[3])
        return output[3]

	# check for OK
    def is_ok(self,output_string):
        if output_string[10:12] == 'OK':
            return True
        else:
            return False

	#RDD
    def read_values(self):
        print("reading values")
        pencil = "{" + str(self.id) + str(self.rs485_addr) + "RDD}\r\n"
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

        split_output = output.split(';')
        # for i in range(0,len(split_output)):
        #     print(str(i) + ": " + split_output[i])
        print("ID_Address_Cmd: " + split_output[0][3:9])
        print("Humidity: " + split_output[1] + "%")
        print("Temperature: " + split_output[5] + " 'C")
        print("Firmware Version: " + split_output[15])
        print("Serial Number: " + split_output[16])
        print("Model: " + split_output[17])

	#REN
    #rs485 address can be 0:63
    def change_rs485_addr(self, new_rs485_addr):
        print("changing RS485 address")
        pencil = "{" + str(self.id) + str(self.rs485_addr) + "REN " + str(self.serial_num) + ";" + str(new_rs485_addr) + ";}\r\n"
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

        if self.is_ok(output):
            print("address change successful!")
            self.rs485_addr = new_rs485_addr
        else:
            print("address failed to change!")

	#HCA
	#calibrate humidity or temperature, hum_or_temp = 1 to change humididty, hum_or_temp = 2 to change temperature
	#ref_val is the referance value for calibration
    def cal_hum_temp(self, hum_or_temp, ref_val):
        timeout_save = self.serial_port.timeout
        self.serial_port.timeout = 3
        print("saving adjustment")
        pencil = "{" + str(self.id) + str(self.rs485_addr) + "HCA 0;" + str(hum_or_temp) + ";0;" + str(ref_val) + ";}\r\n"
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

        if self.is_ok(output) :
            print("adjusting")
            pencil = "{" + str(self.id) + str(self.rs485_addr) + "HCA 0;" + str(hum_or_temp) + ";1;;}\r\n"
            self.serial_port.write(pencil.encode())
            output = str(self.serial_port.readline())

            print(output)

            if self.is_ok(output) :
                print("deleting calibration points")
                pencil = "{"  + str(self.id) + str(self.rs485_addr) + "HCA 0;" + str(hum_or_temp) + ";3;;}\r\n"
                self.serial_port.write(pencil.encode())
                output = str(self.serial_port.readline())

                print(output)

                if self.is_ok(output) :
                    print("calibration is complete")
                else :
                    return print("calibration failed to delete calibration points")
            else :
                return print("calibration failed to adjust")
        else :
            return print("calibration failed to save adjustment")
        self.serial_port.timeout = timeout_save

	# hum_or_temp = 1 to change humididty, hum_or_temp = 2 to change temperature
    def factory_reset_cal(self, hum_or_temp) :
        timeout_save = self.serial_port.timeout
        self.serial_port.timeout = 3
        print("reseting to factory default calibration")
        pencil = "{"  + str(self.id) + str(self.rs485_addr) + "HCA 0;" + str(hum_or_temp) + ";2;;}\r\n"
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

        if self.is_ok(output) :
            return print("factory reset calibration successful!")
        else :
            return print("factory reset calibration failed!")
        self.serial_port.timeout = timeout_save

	#LGC
	#show status of data recording
    def data_rec_status(self) :
        print("showing data recording status")
        pencil = "{"  + str(self.id) + str(self.rs485_addr) + "LGC}\r\n"
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)
		#finish this when you know what the output string will look like
        split_output = output.split(';')
        # for i in range(0,len(split_output)):
        #     print(str(i) + ": " + split_output[i])

        #Recording status
        # 0 = not recording data, 1= recording data
        # Loop recording only: 2 = recording data and memory full (see note below)
        # Loop recording only: 3 = not recording data and memory full (see note below)
        recording_status = split_output[0][11:13]
        if int(recording_status) == 0 :
            recording_status = "not recording data"
        elif int(recording_status) == 1 :
            recording_status = "recording data"
        elif int(recording_status) == 2 :
            recording_status = "recording data and memory full"
        elif int(recording_status) == 3:
            recording_status = "not recording data and memory full"

        #Recording Mode
        # 1 = Start-Stop mode (record until memory is full)
        # 2 = Loop recording (dump oldest data when memory is full and keep recording)
        recording_mode = split_output[1]
        if int(recording_mode) == 1 :
            recording_mode = "Start-Stop mode (record until memory is full)"
        elif int(recording_mode) == 2 :
            recording_mode = "Loop recording (dump oldest data when memory is full and keep recording)"

        #Log Interval
        # Log interval (in multiples of 5 seconds) 1 = 5s, 2 = 10s, etc.
        log_interval = int(split_output[2])
        log_interval = log_interval*5

        #Date & time of first data sample
        # Date and time of the first data sample referenced to Jan, 01, 2000 and expressed in
        # increments of 5 seconds – see note 2 below
        start_time = "date & time : " + ac3k_datetime.decode_datetime(int(split_output[3]))


        #Number of data records
        data_amount = split_output[4]

        print("ID_Address_Cmd: " + split_output[0][3:9])
        print("recording data?: " + recording_status)
        print("recording mode: " + recording_mode)
        print("data logging interval: " + str(log_interval) + " seconds")
        print(start_time)
        print("Amount of data points recorded: " + str(data_amount))


	#stop or start recording data
	#to start/stop recording: startstop_or_loop = 1
	#to loop recording: startstop_or_loop = 2
	#log_interval in values of 5s. ex: 1 = 5s, 2 = 10s...
    def record(self, stop_start, startstop_or_loop, log_interval) :
        print("setting data recording options")
		#set_time is in 5 second intervals since start of 1 Jan. 2000

        set_time = ac3k_datetime.encode_datetime()
        pencil = "{"  + str(self.id) + str(self.rs485_addr) + "LGC " + str(stop_start) + ";" + str(startstop_or_loop) + ";" + str(log_interval) + ";" + str(int(set_time)) + ";" + "}\r\n"

        print(pencil)

        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

        if self.is_ok(output) :
            return print("data recording options updated")
        else :
            return print("data recording options failed to update, please check options")

	#ERD
	#bytes_to_read can be from 0 to 65535
	#each data sample uses 3 bytes
    def download_data(self, output_filename) :
        print("downloading data")
        #Date & time of last data sample
        end_time = time.localtime(time.time())

		# 1) Use the LGC command to read the recording status of the HC2. Note or capture the following:
		# - Recording mode
		# - Log interval (multiple of 5 seconds)
		# - Recording start date and time (see note below)
		# - Number of recorded data samples (in the loop recording mode, disregard the information returned by the
		# LGC command whenever the LGC command reports that the memory is full. In this situation, use 2000
		# as the number of data samples)
        pencil = "{"  + str(self.id) + str(self.rs485_addr) + "LGC}\r\n"
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

        split_output = output.split(';')

        #Recording status
        recording_status = split_output[0][11:13]
        if int(recording_status) == 0 :
            recording_status = "not recording data"
        elif int(recording_status) == 1 :
            recording_status = "recording data"
        elif int(recording_status) == 2 :
            recording_status = "recording data and memory full"
        elif int(recording_status) == 3:
            recording_status = "not recording data and memory full"
        #Recording Mode
        recording_mode = split_output[1]
        if int(recording_mode) == 1 :
            recording_mode = "Start-Stop mode (record until memory is full)"
        elif int(recording_mode) == 2 :
            recording_mode = "Loop recording (dump oldest data when memory is full and keep recording)"
        #Log Interval
        raw_log_interval = int(split_output[2])
        log_interval = raw_log_interval*5
        #Date & time of first data sample
        start_time_encoded = int(split_output[3])
        start_time = ac3k_datetime.decode_datetime(int(split_output[3]))
        #Number of data records
        data_amount = split_output[4]
        data_amount = int(data_amount)
        #max memory is 2000 values
        if data_amount > 2000:
            data_amount = 2000



		#2)
		# Address 2176 is always the address of the first byte of the first recorded data sample. The above example
		# assumes that 2 data samples were recorded. The number of bytes to be downloaded (0006) is equal to 3
        #  times the number of recorded data samples (each data sample uses 3 Bytes)


        #bytes_to_read should be four digits like: 0006
        #bytes_to_read should be assigend from LGC
        bytes_to_read = str(int(data_amount)*3).zfill(4)
        pencil = "{"  + str(self.id) + str(self.rs485_addr) + "ERD 0;2176;" + str(bytes_to_read) + "}\r\n"
        print(pencil)
        self.serial_port.write(pencil.encode())
        output = str(self.serial_port.readline())

        print(output)

		#iterate through a split up output string and perform calculations to get the data
		#output this data to a file or server or something else

        output = output[10:]
        split_output = output.split(';')
        del split_output[-1]

        hum = []
        temp = []

        # for i in range(len(split_output)):
        #     print(split_output[i])
        #     print(i)
        # print(len(split_output))
        for i in range(0, len(split_output), 3) :
            # print(i)
            # print(str(split_output[i]) + ";" + str(split_output[i+1]) + ";" + str(split_output[i+2]))
            value = int(split_output[i]) + 256*int(split_output[i+1]) + 65536*int(split_output[i+2])
            value = float(value)
            hum.append( (value%1024)/10 )
            temp.append(round(math.floor(value/1024) /20-100, 2))

        #saving data to txt file
        # output_file = open('/home/pi/Desktop/'+str(output_filename)+".txt", "w")
        # output_file.write("date & time (format: dd/mm/yyyy hh:mm:ss)\n")
        # output_file.write("Start time:" + start_time+"\n")
        # output_file.write("note: start time should only be used for Start-Stop mode or if loop mode is not at max data collected\n")
        # end_time_string = "date & time (format: dd/mm/yyyy hh:mm:ss) : " + str(end_time.tm_mday) + "/" + str(end_time.tm_mon) + "/" + str(end_time.tm_year) + " " + str(end_time.tm_hour) + ":" + str(end_time.tm_min) + ":" + str(end_time.tm_sec)
        # output_file.write("\nEnd time: " + end_time_string+"\n")
        # output_file.write("note: end time should only be used if loop mode is at max data collected. End time is the time that ERD is called")
        # output_file.write("\nData:\n")
        # for i in range(len(hum)) :
        #     output_file.write(str(i)+". "+str(hum[i])+"%rh    "+str(temp[i])+"'C\n")
        # output_file.close()

        #saving file as CSV
        #time-stamp only works for start-stop mode currently
        with open('/home/pi/Desktop/'+str(output_filename)+".csv", mode ="w") as csv_file:
            fieldnames = ['Time (dd/mm/yyyy hh:mm:ss)','Humidity (%rh)','Temperature (\'C)']
            writer = csv.DictWriter(csv_file, fieldnames=fieldnames)

            writer.writeheader()
            for i in range(len(hum)) :
                data_time_encoded = start_time_encoded + i*raw_log_interval
                data_time = ac3k_datetime.decode_datetime(int(data_time_encoded))
                writer.writerow({'Time (dd/mm/yyyy hh:mm:ss)': data_time,'Humidity (%rh)': str(hum[i]),'Temperature (\'C)':str(temp[i])})

        print("Your data has been saved to " + str(output_filename) + '.csv')

File: HC2_lib_Arduino/HC2/HC2.cpp
//Enoch Chau 2019
#include <Arduino.h>
#include <SoftwareSerial.h>
#include "HC2.h"
#include "ac3k_datetime.h"
#include "RTClib.h"

//assign serial port & rtc
HC2::HC2(SoftwareSerial * serial_port){
  _serial_port = serial_port;
}

//used to init
String HC2::write_rdd(char write_id, String write_rs485 ){
  _serial_port->print("{");
  _serial_port->print(write_id);
  _serial_port->print(write_rs485);
  _serial_port->println("RDD}");

  String output = _serial_port->readString();

  return output;
}

//initlization of id, rs485 address, and serial number
bool HC2::init(){
  bool init_pass = false;

  String output = write_rdd('F',"99");

  if (output != ""){
    init_pass=true;
  }
  // Serial.println(output);
  _id = output.charAt(1);
  _rs485_addr = output.substring(2,4);
  _serial_num = output.substring(75,85);

  Serial.print("ID: "); Serial.println(_id);
  Serial.print("RS485_address: "); Serial.println(_rs485_addr);
  Serial.print("Serial Number: "); Serial.println(_serial_num);
  return init_pass;
}

bool HC2::is_ok(String output_string){
  bool check = false;
  Serial.println(output_string.substring(8,10));
  if (output_string.substring(8,10) == "OK"){
    check = true;
  }
  return check;
}

//read temp and humidity
//RDD
void HC2::read_values(){
  // Serial.println("reading values...");
  String output = write_rdd(_id,_rs485_addr);

  // Serial.println(output);

  String hum = output.substring(13,18);
  String temp = output.substring(30,35);
  String fware_version = output.substring(68,74);
  String model = output.substring(86,89);
  Serial.print("ID: "); Serial.println(_id);
  Serial.print("RS485_address: "); Serial.println(_rs485_addr);
  Serial.print("Serial Number: "); Serial.println(_serial_num);
  Serial.print("Humidity: "); Serial.print(hum);Serial.println("%rh");
  Serial.print("Temperature: "); Serial.print(temp);Serial.println("'C");
  Serial.print("Firmware Version: "); Serial.println(fware_version);
  Serial.print("Model: "); Serial.println(model);
}

//REN
//change RS485 address, addresses can be 0:63
//address must be a two digit string, example: "01", "06", "62"
bool HC2::change_rs485_addr(String new_rs485_addr){
  bool ok = false;

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->print("REN ");
  _serial_port->print(_serial_num);
  _serial_port->print(";");
  _serial_port->print(new_rs485_addr);
  _serial_port->println(";}");

  String output = _serial_port->readString();

  // Serial.println(output);
  ok = is_ok(output);
  if (ok){
    _rs485_addr = new_rs485_addr;
  }
  return(ok);
}

// HCA
// #calibrate humidity or temperature, hum_or_temp = 1 to change humididty, hum_or_temp = 2 to change temperature
// #ref_val is the referance value for calibration
//must ask user for a refereance value for adjustment
//ref value can be from -50 to 200
bool HC2::cal_hum_temp(uint8_t hum_or_temp, int16_t ref_val){
  bool ok = false;
  Serial.println("Saving Adjustment...");

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->print("HCA 0;");
  _serial_port->print(hum_or_temp);
  _serial_port->print(";0;");
  _serial_port->print(ref_val);
  _serial_port->println(";}");

  String output = _serial_port->readString();

  Serial.println(output);

  if (is_ok(output)){
    delay(20);
    Serial.println("Adjusting...");

    _serial_port->print("{");
    _serial_port->print(_id);
    _serial_port->print(_rs485_addr);
    _serial_port->print("HCA 0;");
    _serial_port->print(hum_or_temp);
    _serial_port->println(";1;;}");

    delay(30);
    output = _serial_port->readString();

    Serial.println(output);

    if (is_ok(output)){
      delay(20);
      Serial.println("Deleting Calibration Points...");

      _serial_port->print("{");
      _serial_port->print(_id);
      _serial_port->print(_rs485_addr);
      _serial_port->print("HCA 0;");
      _serial_port->print(hum_or_temp);
      _serial_port->println(";3;;}");

      output = _serial_port->readString();

      Serial.println(output);

      if (is_ok(output)){
        delay(20);
        // Serial.println("Calibration succesful");
        ok = true;
        return(ok);
      }

      else{
        Serial.println("Calibration failed to delete calibration points");
        return(ok);
      }
    }
    else{
      Serial.println("Calibration failed to adjust");
      return(ok);
    }
  }
  else{
    Serial.println("Calibration failed to save adjustment");
    return(ok);
  }
}

//	# hum_or_temp = 1 to change humididty, hum_or_temp = 2 to change temperature
bool HC2::factory_reset_cal(uint8_t hum_or_temp){
  bool ok = false;
  // Serial.println("Resetting to factory default calibration...");

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->print("HCA 0;");
  _serial_port->print(hum_or_temp);
  _serial_port->println(";2;;}");

  String output = _serial_port->readString();

  // Serial.println(output);

  delay(20);
  ok = is_ok(output);
  return(ok);
}

//LGC
//show status of data recording
void HC2::data_rec_status(){
  Serial.println("showing data recording status.");

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->println("LGC}");

  String output = _serial_port->readString();

  Serial.println(output);

  // #Recording status
  //     # 0 = not recording data, 1= recording data
  //     # Loop recording only: 2 = recording data and memory full (see note below)
  //     # Loop recording only: 3 = not recording data and memory full (see note below)
  String recording_status = "";
  if (output[10] == '0'){
    recording_status = "not recording data";
  }
  else if (output[10] == '1'){
    recording_status = "recording data";
  }
  else if (output[10] == '2'){
    recording_status = "recording data and memory full";
  }
  else if (output[10] == '3'){
    recording_status = "not recording data and memory full";
  }

  // #Recording Mode
  //     # 1 = Start-Stop mode (record until memory is full)
  //     # 2 = Loop recording (dump oldest data when memory is full and keep recording)
  String recording_mode = "";
  if (output[14] == '1'){
    recording_mode = "Start-Stop mode (record until memory is full)";
  }
  else if (output[14] == '2'){
    recording_mode = "Loop recording (dump oldest data when memory is full and keep recording)";
  }

  // #Log Interval
  // # Log interval (in multiples of 5 seconds) 1 = 5s, 2 = 10s, etc.

  // String log_interval_str = output.substring(16,21);
  // int log_interval_int = log_interval_str.toInt();
  // log_interval_int = log_interval_int*5;
  int log_interval_int = output.substring(16,21).toInt();
  log_interval_int = log_interval_int*5;


  // #Date & time of first data sample
  // # Date and time of the first data sample referenced to Jan, 01, 2000 and expressed in
  // # increments of 5 seconds – see note 2 below
  DateTime start_time = decode_datetime(output.substring(22,32).toInt());

  Serial.print("start time: ");
  Serial.println(start_time.timestamp());

  Serial.print("recording data?: ");
  Serial.println(recording_status);

  Serial.print("recording mode: ");
  Serial.println(recording_mode);

  Serial.print("data logging interval: ");
  Serial.print(log_interval_int);
  Serial.println(" sec.");

  Serial.print("Amount of data points recorded: ");
  Serial.println(output.substring(33,38));
}

// #stop or start recording data
// #to start/stop recording mode: startstop_or_loop = 1
// #to loop recording mode: startstop_or_loop = 2
// #log_interval in values of 5s. ex: 1 = 5s, 2 = 10s...
void HC2::record(bool stop_start, uint8_t startstop_or_loop, uint16_t log_interval, uint32_t set_time){
  Serial.println("setting data recording options");

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->print("LGC ");
  _serial_port->print(stop_start);
  _serial_port->print(";");
  _serial_port->print(startstop_or_loop);
  _serial_port->print(";");
  _serial_port->print(log_interval);
  _serial_port->print(";");
  _serial_port->print(set_time);
  _serial_port->println(";}");

  String output = _serial_port->readString();

  Serial.println(output);

  if (is_ok(output)){
    Serial.println("Data recording options updated!");
  }
  else{
    Serial.println("Data recording options failed to update!");
  }
}

// #ERD
// #bytes_to_read can be from 0 to 65535
// #each data sample uses 3 bytes
void HC2::download_data(String output_filename, uint32_t end_time){
  Serial.println("Downloading Data");

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->println("LGC}");

  Serial.println("LGC done");

  String output = _serial_port->readString();

  Serial.println(output);

//start time
  DateTime start_time = decode_datetime(output.substring(22,32).toInt());

  //log interval, raw log interval in increments of 5sec
  int log_interval_int = output.substring(16,21).toInt();

//Number of data records; max memory is 2000 values
// #bytes_to_read should be four digits like: 0006
// #bytes_to_read should be assigend from LGC
  int bytes_to_read = output.substring(33,38).toInt() * 3;

  Serial.println("starting erd");

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->print("ERD 0;2176;");
  if (bytes_to_read < 1000){ //zero padding
    _serial_port->print(0);
  };
  if (bytes_to_read < 100){
    _serial_port->print(0);
  };
  if (bytes_to_read < 10){
    _serial_port->print(0);
  };
  _serial_port->print(bytes_to_read);
  _serial_port->println("}");

  Serial.println("erd done");
  delay(3000);
  //each byte_to_read is 4 chars// ex: '038;'
  //the header and checksum are 9 chars
  output = "";
  while(_serial_port->available()>0){
    output += (char)_serial_port->read();
    delay(10);
  }


  // output = _serial_port->readString();
  //
  Serial.println(output);

  //remove the header
  //do calculations to get hum and temp
  //export that data
}

//--------------------------------------------------------------
//Functions that return raw data stream from HC2

//read temp and humidity
//RDD

String HC2::read_values_ds(){
  // Serial.println("reading values...");
  String output = write_rdd(_id,_rs485_addr);
  // Serial.println(output);
  return(output);
}

void HC2::decode_rdd_ds(String output){
  String hum = output.substring(13,18);
  String temp = output.substring(30,35);
  String fware_version = output.substring(68,74);
  String model = output.substring(86,89);
  Serial.print("ID: "); Serial.println(_id);
  Serial.print("RS485_address: "); Serial.println(_rs485_addr);
  Serial.print("Serial Number: "); Serial.println(_serial_num);
  Serial.print("Humidity: "); Serial.print(hum);Serial.println("%rh");
  Serial.print("Temperature: "); Serial.print(temp);Serial.println("'C");
  Serial.print("Firmware Version: "); Serial.println(fware_version);
  Serial.print("Model: "); Serial.println(model);
}
//LGC
String HC2::data_rec_status_ds(){

  _serial_port->print("{");
  _serial_port->print(_id);
  _serial_port->print(_rs485_addr);
  _serial_port->println("LGC}");

  String output = _serial_port->readString();

  Serial.println(output);
  return(output);
}

void decode_lgc_ds(String output){
  String recording_status = "";
  if (output[10] == '0'){
    recording_status = "not recording data";
  }
  else if (output[10] == '1'){
    recording_status = "recording data";
  }
  else if (output[10] == '2'){
    recording_status = "recording data and memory full";
  }
  else if (output[10] == '3'){
    recording_status = "not recording data and memory full";
  }
  // #Recording Mode
  //     # 1 = Start-Stop mode (record until memory is full)
  //     # 2 = Loop recording (dump oldest data when memory is full and keep recording)
  String recording_mode = "";
  if (output[14] == '1'){
    recording_mode = "Start-Stop mode (record until memory is full)";
  }
  else if (output[14] == '2'){
    recording_mode = "Loop recording (dump oldest data when memory is full and keep recording)";
  }

  // #Log Interval
  // # Log interval (in multiples of 5 seconds) 1 = 5s, 2 = 10s, etc.

  // String log_interval_str = output.substring(16,21);
  // int log_interval_int = log_interval_str.toInt();
  // log_interval_int = log_interval_int*5;
  int log_interval_int = output.substring(16,21).toInt();
  log_interval_int = log_interval_int*5;


  // #Date & time of first data sample
  // # Date and time of the first data sample referenced to Jan, 01, 2000 and expressed in
  // # increments of 5 seconds – see note 2 below
  DateTime start_time = decode_datetime(output.substring(22,32).toInt());

  Serial.print("start time: ");
  Serial.println(start_time.timestamp());

  Serial.print("recording data?: ");
  Serial.println(recording_status);

  Serial.print("recording mode: ");
  Serial.println(recording_mode);

  Serial.print("data logging interval: ");
  Serial.print(log_interval_int);
  Serial.println(" sec.");

  Serial.print("Amount of data points recorded: ");
  Serial.println(output.substring(33,38));
}
